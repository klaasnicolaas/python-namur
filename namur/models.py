"""Models for Open Data Platform of Namur."""
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime
from typing import Any

import pytz


@dataclass
class ParkingSpot:
    """Object representing a parking spot."""

    spot_id: str
    parking_type: str
    street: str
    longitude: float
    latitude: float
    created_at: datetime
    updated_at: datetime

    @classmethod
    def from_json(cls: type[ParkingSpot], data: dict[str, Any]) -> ParkingSpot:
        """Return a ParkingSpot object from a JSON dictionary.

        Args:
        ----
            data: The JSON data from the API.

        Returns:
        -------
            A ParkingSpot object.
        """
        attr = data["fields"]
        geo = data["geometry"]["coordinates"]
        return cls(
            spot_id=attr.get("identifiant"),
            parking_type=attr.get("type_parking"),
            street=attr.get("rue_nom"),
            longitude=geo[0],
            latitude=geo[1],
            created_at=datetime.strptime(
                attr.get("date_creation"),
                "%Y-%m-%d",
            ).astimezone(pytz.utc),
            updated_at=datetime.strptime(
                data["record_timestamp"],
                "%Y-%m-%dT%H:%M:%SZ",
            ).astimezone(pytz.utc),
        )
