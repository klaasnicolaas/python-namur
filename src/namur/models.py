"""Models for Open Data Platform of Namur."""

from __future__ import annotations

from dataclasses import dataclass
from datetime import UTC, datetime
from enum import Enum
from typing import Any


class ParkingType(str, Enum):
    """Enum representing the type of parking."""

    NORMAL = "Place normale"
    GARAGE = "Devant accès/garage"
    PMR = "PMR"
    DROP_OFF = "Dépose-minute"
    DELIVERY = "Livraison"
    POLICE = "Police"
    TAXI = "Taxi"
    CAR_SHARING = "Car-sharing"
    RECYCLING = "Recyclage"
    CAR = "Car"
    SCHOOL_BUS = "Bus scolaire"
    ELECTRIC_CAR = "Borne électrique"
    RESERVED = "Réservé"


@dataclass
class ParkingSpot:
    """Object representing a parking spot."""

    spot_id: str
    parking_type: str
    street: str
    longitude: float
    latitude: float
    created_at: datetime
    updated_at: datetime

    @classmethod
    def from_json(cls: type[ParkingSpot], data: dict[str, Any]) -> ParkingSpot:
        """Return a ParkingSpot object from a JSON dictionary.

        Args:
        ----
            data: The JSON data from the API.

        Returns:
        -------
            A ParkingSpot object.

        """
        attr = data["fields"]
        geo = data["geometry"]["coordinates"]
        return cls(
            spot_id=attr.get("identifiant"),
            parking_type=attr.get("type_parking"),
            street=attr.get("rue_nom"),
            longitude=geo[0],
            latitude=geo[1],
            created_at=datetime.strptime(
                attr.get("date_creation"),
                "%Y-%m-%d",
            ).replace(tzinfo=UTC),
            updated_at=datetime.strptime(
                data["record_timestamp"],
                "%Y-%m-%dT%H:%M:%SZ",
            ).replace(tzinfo=UTC),
        )
