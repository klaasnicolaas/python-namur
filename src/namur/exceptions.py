"""Asynchronous Python client providing Open Data information of Namur."""


class ODPNamurError(Exception):
    """Generic Open Data Platform Namur exception."""


class ODPNamurConnectionError(ODPNamurError):
    """Open Data Platform Namur - connection exception."""


class ODPNamurTypeError(ODPNamurError):
    """Open Data Platform Namur - parking type exception."""


class ODPNamurResultsError(ODPNamurError):
    """Open Data Platform Namur - no results exception."""
