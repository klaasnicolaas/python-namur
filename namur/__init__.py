"""Asynchronous Python client providing Open Data information of Namur."""

from .exceptions import (
    ODPNamurConnectionError,
    ODPNamurError,
    ODPNamurResultsError,
    ODPNamurTypeError,
)
from .models import ParkingSpot
from .namur import ODPNamur

__all__ = [
    "ODPNamur",
    "ODPNamurConnectionError",
    "ODPNamurError",
    "ODPNamurResultsError",
    "ODPNamurTypeError",
    "ParkingSpot",
]
