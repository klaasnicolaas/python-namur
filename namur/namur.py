"""Asynchronous Python client providing Open Data information of Namur."""
from __future__ import annotations

import asyncio
import socket
from dataclasses import dataclass
from importlib import metadata
from typing import Any

import aiohttp
import async_timeout
from aiohttp import hdrs
from yarl import URL

from .exceptions import (
    ODPNamurConnectionError,
    ODPNamurError,
    ODPNamurResultsError,
    ODPNamurTypeError,
)
from .models import ParkingSpot


@dataclass
class ODPNamur:
    """Main class for handling data fetchting from Open Data Platform of Namur."""

    request_timeout: float = 10.0
    session: aiohttp.client.ClientSession | None = None

    _close_session: bool = False

    @staticmethod
    async def define_type(parking_type: int) -> str:
        """Define the parking type.

        Args:
            parking_type: The selected parking type number.

        Returns:
            The parking type as string.

        Raises:
            ODPNamurTypeError: If the parking type is not listed.
        """
        options = {
            1: "Place normale",
            2: "Devant accès/garage",
            3: "PMR",
            4: "Dépose-minute",
            5: "Livraison",
            6: "Police",
            7: "Taxi",
            8: "Car-sharing",
            9: "Recyclage",
            10: "Car",
            11: "Bus scolaire",
            12: "Borne électrique",
            13: "Réservé",
        }.get(parking_type)

        # Check if the parking type is listed
        if options is None:
            raise ODPNamurTypeError(
                "The selected number does not match the list of parking types"
            )
        return options

    async def _request(
        self,
        uri: str,
        *,
        method: str = hdrs.METH_GET,
        params: dict[str, Any] | None = None,
    ) -> Any:
        """Handle a request to the Open Data Platform API of Namur.

        Args:
            uri: Request URI, without '/', for example, 'status'
            method: HTTP method to use, for example, 'GET'
            params: Extra options to improve or limit the response.

        Returns:
            A Python dictionary (json) with the response from
            the Open Data Platform API of Namur.

        Raises:
            ODPNamurConnectionError: An error occurred while
                communicating with the Open Data Platform API.
            ODPNamurError: Received an unexpected response from
                the Open Data Platform API.
        """
        version = metadata.version(__package__)
        url = URL.build(
            scheme="https", host="data.namur.be", path="/api/records/1.0/"
        ).join(URL(uri))

        headers = {
            "Accept": "application/json, text/plain",
            "User-Agent": f"PythonNamur/{version}",
        }

        if self.session is None:
            self.session = aiohttp.ClientSession()
            self._close_session = True

        try:
            async with async_timeout.timeout(self.request_timeout):
                response = await self.session.request(
                    method,
                    url,
                    params=params,
                    headers=headers,
                    ssl=True,
                )
                response.raise_for_status()
        except asyncio.TimeoutError as exception:
            raise ODPNamurConnectionError(
                "Timeout occurred while connecting to the Open Data Platform API."
            ) from exception
        except (aiohttp.ClientError, socket.gaierror) as exception:
            raise ODPNamurConnectionError(
                "Error occurred while communicating with the Open Data Platform API."
            ) from exception

        content_type = response.headers.get("Content-Type", "")
        if "application/json" not in content_type:
            text = await response.text()
            raise ODPNamurError(
                "Unexpected content type response from the Open Data Platform API",
                {"Content-Type": content_type, "response": text},
            )

        return await response.json()

    async def parking_spaces(
        self, limit: int = 10, parking_type: int = 1
    ) -> list[ParkingSpot]:
        """Get all the parking locations.

        Args:
            limit: Number of rows to return.
            parking_type: The selected parking type number.

        Returns:
            A list of ParkingSpot objects.

        Raises:
            ODPNamurResultsError: When no results are found.
        """
        results: list[ParkingSpot] = []
        locations = await self._request(
            "search/",
            params={
                "dataset": "namur-parking-emplacements",
                "rows": limit,
                "refine.type_parking": await self.define_type(parking_type),
            },
        )

        for item in locations["records"]:
            results.append(ParkingSpot.from_json(item))
        if not results:
            raise ODPNamurResultsError("No parking locations were found")
        return results

    async def close(self) -> None:
        """Close open client session."""
        if self.session and self._close_session:
            await self.session.close()

    async def __aenter__(self) -> ODPNamur:
        """Async enter.

        Returns:
            The Open Data Platform Namur object.
        """
        return self

    async def __aexit__(self, *_exc_info: Any) -> None:
        """Async exit.

        Args:
            _exc_info: Exec type.
        """
        await self.close()
