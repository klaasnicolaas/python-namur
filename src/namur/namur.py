"""Asynchronous Python client providing Open Data information of Namur."""
from __future__ import annotations

import asyncio
import socket
from dataclasses import dataclass
from importlib import metadata
from typing import Any, Self, cast

from aiohttp import ClientError, ClientSession
from aiohttp.hdrs import METH_GET
from yarl import URL

from .exceptions import (
    ODPNamurConnectionError,
    ODPNamurError,
    ODPNamurResultsError,
    ODPNamurTypeError,
)
from .models import ParkingSpot


@dataclass
class ODPNamur:
    """Main class for handling data fetchting from Open Data Platform of Namur."""

    request_timeout: float = 10.0
    session: ClientSession | None = None

    _close_session: bool = False

    @staticmethod
    async def define_type(parking_type: int) -> str:
        """Define the parking type.

        Args:
        ----
            parking_type: The selected parking type number.

        Returns:
        -------
            The parking type as string.

        Raises:
        ------
            ODPNamurTypeError: If the parking type is not listed.
        """
        options = {
            1: "Place normale",
            2: "Devant accès/garage",
            3: "PMR",
            4: "Dépose-minute",
            5: "Livraison",
            6: "Police",
            7: "Taxi",
            8: "Car-sharing",
            9: "Recyclage",
            10: "Car",
            11: "Bus scolaire",
            12: "Borne électrique",
            13: "Réservé",
        }.get(parking_type)

        # Check if the parking type is listed
        if options is None:
            msg = "The selected number does not match the list of parking types"
            raise ODPNamurTypeError(
                msg,
            )
        return options

    async def _request(
        self,
        uri: str,
        *,
        method: str = METH_GET,
        params: dict[str, Any] | None = None,
    ) -> Any:
        """Handle a request to the Open Data Platform API of Namur.

        Args:
        ----
            uri: Request URI, without '/', for example, 'status'
            method: HTTP method to use, for example, 'GET'
            params: Extra options to improve or limit the response.

        Returns:
        -------
            A Python dictionary (json) with the response from
            the Open Data Platform API of Namur.

        Raises:
        ------
            ODPNamurConnectionError: An error occurred while
                communicating with the Open Data Platform API.
            ODPNamurError: Received an unexpected response from
                the Open Data Platform API.
        """
        version = metadata.version(__package__)
        url = URL.build(
            scheme="https",
            host="data.namur.be",
            path="/api/records/1.0/",
        ).join(URL(uri))

        headers = {
            "Accept": "application/json, text/plain",
            "User-Agent": f"PythonNamur/{version}",
        }

        if self.session is None:
            self.session = ClientSession()
            self._close_session = True

        try:
            async with asyncio.timeout(self.request_timeout):
                response = await self.session.request(
                    method,
                    url,
                    params=params,
                    headers=headers,
                    ssl=True,
                )
                response.raise_for_status()
        except asyncio.TimeoutError as exception:
            msg = "Timeout occurred while connecting to the Open Data Platform API."
            raise ODPNamurConnectionError(
                msg,
            ) from exception
        except (ClientError, socket.gaierror) as exception:
            msg = "Error occurred while communicating with the Open Data Platform API."
            raise ODPNamurConnectionError(
                msg,
            ) from exception

        content_type = response.headers.get("Content-Type", "")
        if "application/json" not in content_type:
            text = await response.text()
            msg = "Unexpected content type response from the Open Data Platform API"
            raise ODPNamurError(
                msg,
                {"Content-Type": content_type, "response": text},
            )

        return cast(dict[str, Any], await response.json())

    async def parking_spaces(
        self,
        limit: int = 10,
        parking_type: int = 1,
    ) -> list[ParkingSpot]:
        """Get all the parking locations.

        Args:
        ----
            limit: Number of rows to return.
            parking_type: The selected parking type number.

        Returns:
        -------
            A list of ParkingSpot objects.

        Raises:
        ------
            ODPNamurResultsError: When no results are found.
        """
        locations = await self._request(
            "search/",
            params={
                "dataset": "namur-parking-emplacements",
                "rows": limit,
                "refine.type_parking": await self.define_type(parking_type),
            },
        )

        results: list[ParkingSpot] = [
            ParkingSpot.from_json(item) for item in locations["records"]
        ]
        if not results:
            msg = "No parking locations were found"
            raise ODPNamurResultsError(msg)
        return results

    async def close(self) -> None:
        """Close open client session."""
        if self.session and self._close_session:
            await self.session.close()

    async def __aenter__(self) -> Self:
        """Async enter.

        Returns
        -------
            The Open Data Platform Namur object.
        """
        return self

    async def __aexit__(self, *_exc_info: object) -> None:
        """Async exit.

        Args:
        ----
            _exc_info: Exec type.
        """
        await self.close()
