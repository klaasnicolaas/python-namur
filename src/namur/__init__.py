"""Asynchronous Python client providing Open Data information of Namur."""

from .exceptions import (
    ODPNamurConnectionError,
    ODPNamurError,
    ODPNamurResultsError,
)
from .models import ParkingSpot, ParkingType
from .namur import ODPNamur

__all__ = [
    "ODPNamur",
    "ODPNamurConnectionError",
    "ODPNamurError",
    "ODPNamurResultsError",
    "ParkingSpot",
    "ParkingType",
]
