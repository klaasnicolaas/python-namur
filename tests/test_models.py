"""Test the models."""

from __future__ import annotations

import pytest
from aiohttp import ClientSession
from aresponses import ResponsesMockServer

from namur import ODPNamur, ODPNamurResultsError, ODPNamurTypeError, ParkingSpot

from . import load_fixtures


async def test_parking_model(aresponses: ResponsesMockServer) -> None:
    """Test the parking model."""
    aresponses.add(
        "data.namur.be",
        "/api/records/1.0/search/",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixtures("parking_pmr.json"),
        ),
    )
    async with ClientSession() as session:
        client = ODPNamur(session=session)
        locations: list[ParkingSpot] = await client.parking_spaces(parking_type=1)
        for item in locations:
            assert item.spot_id is not None
            assert item.street is not None
            assert item.updated_at is not None


async def test_no_parking_results(aresponses: ResponsesMockServer) -> None:
    """Test if there are no parking results."""
    aresponses.add(
        "data.namur.be",
        "/api/records/1.0/search/",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixtures("no_parking.json"),
        ),
    )
    async with ClientSession() as session:
        client = ODPNamur(session=session)
        with pytest.raises(ODPNamurResultsError):
            await client.parking_spaces(parking_type=3)


async def test_no_parking_type(aresponses: ResponsesMockServer) -> None:
    """Test when parking_type doesn't exist."""
    aresponses.add(
        "data.namur.be",
        "/api/records/1.0/search/",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixtures("parking_pmr.json"),
        ),
    )
    async with ClientSession() as session:
        client = ODPNamur(session=session)
        with pytest.raises(ODPNamurTypeError):
            await client.parking_spaces(parking_type=20)
