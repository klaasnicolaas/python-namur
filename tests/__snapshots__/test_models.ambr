# serializer version: 1
# name: test_parking_model
  list([
    ParkingSpot(spot_id='2172', parking_type='PMR', street="Place de l'Ecole des Cadets", longitude=4.87520726773486, latitude=50.46580805599107, created_at=datetime.datetime(2020, 9, 15, 22, 0, tzinfo=<UTC>), updated_at=datetime.datetime(2022, 8, 27, 23, 10, tzinfo=<UTC>)),
    ParkingSpot(spot_id='2529', parking_type='PMR', street='Rue Lucien Namêche', longitude=4.868819424458087, latitude=50.46702277846857, created_at=datetime.datetime(2020, 9, 15, 22, 0, tzinfo=<UTC>), updated_at=datetime.datetime(2022, 8, 27, 23, 10, tzinfo=<UTC>)),
    ParkingSpot(spot_id='14378', parking_type='PMR', street='Rue du Treizième de Ligne', longitude=4.821362935779792, latitude=50.47341760702161, created_at=datetime.datetime(2020, 9, 15, 22, 0, tzinfo=<UTC>), updated_at=datetime.datetime(2022, 8, 27, 23, 10, tzinfo=<UTC>)),
    ParkingSpot(spot_id='15582', parking_type='PMR', street='Rue Charles Lamquet', longitude=4.87941976567549, latitude=50.4639856538556, created_at=datetime.datetime(2020, 9, 15, 22, 0, tzinfo=<UTC>), updated_at=datetime.datetime(2022, 8, 27, 23, 10, tzinfo=<UTC>)),
    ParkingSpot(spot_id='15591', parking_type='PMR', street='Place Joséphine Charlotte', longitude=4.877104301130791, latitude=50.46122805654307, created_at=datetime.datetime(2020, 9, 15, 22, 0, tzinfo=<UTC>), updated_at=datetime.datetime(2022, 8, 27, 23, 10, tzinfo=<UTC>)),
    ParkingSpot(spot_id='15586', parking_type='PMR', street='Place Notre-Dame de la Paix', longitude=4.896604222426458, latitude=50.4406149774601, created_at=datetime.datetime(2020, 9, 15, 22, 0, tzinfo=<UTC>), updated_at=datetime.datetime(2022, 8, 27, 23, 10, tzinfo=<UTC>)),
    ParkingSpot(spot_id='15590', parking_type='PMR', street='Chaussée de Liège', longitude=4.88408357144082, latitude=50.45869389215245, created_at=datetime.datetime(2020, 9, 15, 22, 0, tzinfo=<UTC>), updated_at=datetime.datetime(2022, 8, 27, 23, 10, tzinfo=<UTC>)),
    ParkingSpot(spot_id='15592', parking_type='PMR', street='Place Joséphine Charlotte', longitude=4.877126555534669, latitude=50.461210614713515, created_at=datetime.datetime(2020, 9, 15, 22, 0, tzinfo=<UTC>), updated_at=datetime.datetime(2022, 8, 27, 23, 10, tzinfo=<UTC>)),
    ParkingSpot(spot_id='15594', parking_type='PMR', street='Rue Commandant Tilot', longitude=4.876987945585452, latitude=50.4624169952671, created_at=datetime.datetime(2020, 9, 15, 22, 0, tzinfo=<UTC>), updated_at=datetime.datetime(2022, 8, 27, 23, 10, tzinfo=<UTC>)),
    ParkingSpot(spot_id='14175', parking_type='PMR', street='Rue du Bouquet', longitude=4.837716682008543, latitude=50.47477032756567, created_at=datetime.datetime(2020, 9, 15, 22, 0, tzinfo=<UTC>), updated_at=datetime.datetime(2022, 8, 27, 23, 10, tzinfo=<UTC>)),
  ])
# ---
